lib/client.ts

```diff
@@ -0,0 +1,361 @@
+  }
+  circuitBreaker?: (false | {
+    windowDuration?: number,
+    numBuckets?: number,
+    timeoutDuration?: number,
+    errorThreshold?: number,
+    volumeThreshold?: number
+  })
+  defaultRequestOptions?: ServiceClientRequestOptions
+}
+
+/**
+ * Internal only, this interface guarantees, that the service client has all options available at runtime.
+ */
+class ServiceClientStrictOptions {
+  hostname: string
+  filters: ServiceClientRequestFilter[]
+  timing: boolean
+  retryOptions: {
+    retries: number
+    factor: number
+    minTimeout: number
+    maxTimeout: number
+    randomize: boolean
+    shouldRetry: (err?: Error, req?: ServiceClientRequestOptions) => boolean
+    onRetry: (currentAttempt?: number, err?: Error, req?: ServiceClientRequestOptions) => void
+  }
+  defaultRequestOptions: ServiceClientRequestOptions
+
+  constructor(options: ServiceClientOptions){
+    if (!options.hostname) {
+      throw new Error('Please provide a `hostname` for this client')
+    }
+
+    this.hostname = options.hostname
+    this.filters = Array.isArray(options.filters) ? options.filters : [...ServiceClient.DEFAULT_FILTERS]
+    this.timing = Boolean(options.timing)
+    this.retryOptions = Object.assign({
+      retries: 0,
+      factor: 2,
+      minTimeout: 200,
+      maxTimeout: 400,
+      randomize: true,
+      shouldRetry () {
+        return true
+      },
+      onRetry () {}
+    }, options.retryOptions)
+
+    if ((this.retryOptions.minTimeout || 0) > (this.retryOptions.maxTimeout || 0)) {
+      throw new TypeError('The `maxTimeout` must be equal to or greater than the `minTimeout`')
+    }
+
+    this.defaultRequestOptions = Object.assign({
+      protocol: 'https:',
+      pathname: '/',
+      timeout: 2000
+    }, options.defaultRequestOptions)
+  }
+}
+
+/**
+ * A custom error returned in case something goes wrong.
+ */
+export class ServiceClientError extends Error {
+  constructor (originalError: Error, public type: string, public response?: ServiceClientResponse) {
+    super(`${type}. ${originalError.message || ''}`)
+    Object.assign(this, originalError)
+  }
+}
+
+/**
+ * This function takes a response and if it is of type json, tries to parse the body.
+ */
+const decodeResponse = (response: ServiceClientResponse): ServiceClientResponse => {
+  const contentType = response.headers['content-type'] || (response.body ? 'application/json' : '')
+  if (contentType.startsWith('application/json') && typeof response.body !== 'object') {
+    try {
+      response.body = JSON.parse(response.body)
+    } catch (error) {
+      throw new ServiceClientError(
+        error, ServiceClient.BODY_PARSE_FAILED, response
+      )
+    }
+  }
+  return response;
+}
+
+/**
+ * Wrapper that makes sure that all error coming out
+ * of ServiceClients are actual ServiceClientError
+ */
+const wrapFailedError = (type: string, error: Error | ServiceClientError, responseThunk?: Function) => {
```

Comment:
It doesn't add too much, but I would change type from `Function` to `() => any`, which tells that we don't expect any arguments

Reply:
will do